packages/neon-api/src/transaction/validator.ts

```diff
@@ -0,0 +1,298 @@
+import { getNetworkFee, getScriptHashesFromTxWitnesses } from "./util";
+
+export enum ValidationAttributes {
+  None = 0,
+  ValidUntilBlock = 1 << 0,
+  SystemFee = 1 << 1,
+  NetworkFee = 1 << 2,
+  Script = 1 << 3,
+  All = 15
+}
+
+export interface ValidationSuggestion<T> {
+  /**
+   * Whether this is auto-fixed by validator
+   * Validator will try to fix the transaction when param `autoFix` is set to TRUE
+   */
+  fixed: boolean;
+  prev: T;
+  suggestion: T;
+}
+
+export interface ValidationResult {
+  /**
+   * Whether the transaction is valid after validation
+   */
+  valid: boolean;
+  suggestions?: {
+    validaUntilBlock?: ValidationSuggestion<number>;
+    script?: Array<string>;
+    systemFee?: ValidationSuggestion<u.Fixed8>;
+    networkFee?: ValidationSuggestion<u.Fixed8>;
+  };
+}
+
+/**
+ * A class with functions to validate transaction
+ */
+export class TransactionValidator {
+  /**
+   * Transaction will be validated on this rpc node
+   */
+  public rpcClient: rpc.RPCClient;
+
+  /**
+   * Transaction that will be validated
+   */
+  public transaction: tx.Transaction;
+
+  constructor(rpc: rpc.RPCClient, transaction: tx.Transaction) {
+    this.rpcClient = rpc;
+    this.transaction = transaction;
+  }
+
+  /**
+   * validate transaction attribute - validUntilBlock
+   * @param autoFix will automatically fix transaction if specified as true
+   */
+  async validateValidUntilBlock(autoFix = false): Promise<ValidationResult> {
+    const { validUntilBlock } = this.transaction;
+    const height = await this.rpcClient.getBlockCount();
+    if (
+      validUntilBlock <= height ||
+      validUntilBlock >= height + tx.Transaction.MAX_TRANSACTION_LIFESPAN
+    ) {
+      const vub_suggestion =
+        tx.Transaction.MAX_TRANSACTION_LIFESPAN + height - 1;
+      if (autoFix) {
+        this.transaction.validUntilBlock = vub_suggestion;
+        return {
+          valid: true,
+          suggestions: {
+            validaUntilBlock: {
+              fixed: true,
+              prev: validUntilBlock,
+              suggestion: vub_suggestion
+            }
+          }
+        };
+      }
+      return {
+        valid: false,
+        suggestions: {
+          validaUntilBlock: {
+            fixed: false,
+            prev: validUntilBlock,
+            suggestion: vub_suggestion
+          }
+        }
+      };
+    }
+    return {
+      valid: true
+    };
+  }
+
+  private async _validateIntent(
+    intent: sc.ScriptIntent
+  ): Promise<string | undefined> {
+    const { scriptHash, operation } = intent;
+    const manifest = await this.rpcClient.getContractState(scriptHash);
+    if (manifest === null) {
+      return `Unknow contract ${scriptHash}`;
+    }
+
+    if (operation) {
+      if (
+        manifest.abi.methods.map(method => method.name).indexOf(operation) < 0
+      ) {
+        return `Unknow method ${operation} for constract ${scriptHash}`;
+      }
+    }
+
+    return undefined;
```

Comment:
You might want to consider returning empty string or null instead of undefined. Undefined is usually not intentional.